### PR TITLE
Add view all loans command

### LINKED ISSUE #55: Implement command to view all active loans
### Feature description

Command: `viewloan active`

Output: a GUI that displays all currently active loans that have not been returned

### Implementation overview

- Similar to `viewloan` command
- Instead of adding only 1 person's loan, iterate through all contacts in addressbook and add all to `ObservableList` in `Model`
- Sort by some criteria (default: due date)

### PR DESCRIPTION
**Pull request by narwhalsilent**
Fixes #55 and #32.

### Refactor `LoanRecords` as `UniqueLoanList`

- The new class use `UniquePersonList` as a model in implementing methods.

### Extract individual loan lists to global loan lists

- The `UniqueLoanList` is now a field of `AddressBook` as opposed individual persons.
- The class `Loan` now holds the assignee `Person` as a field.

### Refactor the display of loans

Class changes:
- `ModelManager` now holds a `sortedLoanList` of loans wrapped around a `filteredLoanList`.
- This `filteredLoanList`, in turn, wraps around the `uniqueLoanList` of the AddressBook, analogous to `filteredPersonList`.

Implementation changes:
- The `filteredLoanList` controls which loans are shown. Currently there are two options: show all active loans or show active loans attached to a person.
- The `sortedLoanList` controls the order of loans. Currently it is hardcoded that loans due earlier are placed higher.

### Add new command `viewloans`

- With above changes, this command is implemented similarly to `list`. It displays all active loans

### Refactor `viewloan` command

- Instead of getting a personal `LoanRecords`, this command now simply updates the predicate on `filteredLoanList`, similar to `filter` command.

### Refactor `deleteloan` and `markloan` command

- Referring to the index of person is no longer required due to previous change.
- New syntax is `deleteloan/markloan INDEX` where `INDEX` is the index of the loan in the last shown loan list.

### Refactor storage

- The storage system is comprehensively revised to accomodate all changes above.

### Change generic data

- The generic data for when `addressbook.json` does not exist is changed to reflect the above changes.

### Bugs and issues

The following are **essential** changes that are not covered in this PR:

- [ ] Change the command messages of `deleteloan` and `markloan` to reflect the change of parameters.
- [ ] Change the DG to reflect the change of implementation of `deleteloan` and `markloan` methods.
- [ ] Change the automated tests to reflect the refactoring of `Person`, `Loan` and `LoanRecords`.
- [ ] Refactor the generation of `Analytics` in this new implementation.
- [ ] Refactor `EditCommand` related objects to reflect the deletion of `LoanRecords` from `Person`.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Loan.java [hunk 1]
```diff
@@ -22,40 +22,45 @@ public class Loan {
...
+     * @param id         A valid id.
+     * @param value      A valid value.
+     * @param startDate  A valid start date.
      * @param returnDate A valid return date.
+     * @param assignee   A valid assignee.
      */
-    public Loan(int id, float value, Date startDate, Date returnDate) {
-        requireAllNonNull(id, value, startDate, returnDate);
+    public Loan(int id, float value, Date startDate, Date returnDate, Person assignee) {
+        requireAllNonNull(id, value, startDate, returnDate, assignee);
```
Comment (marcus-ny) [inline src/main/java/seedu/address/model/person/Loan.java:37]: Good defensive coding

File: src/main/java/seedu/address/model/person/Loan.java [hunk 2]
```diff
@@ -22,40 +22,45 @@ public class Loan {
...
+     * @param id         A valid id.
+     * @param value      A valid value.
+     * @param startDate  A valid start date.
      * @param returnDate A valid return date.
      * @param isReturned A valid return status.
+     * @param assignee   A valid assignee.
      */
-    public Loan(int id, float value, Date startDate, Date returnDate, boolean isReturned) {
-        requireAllNonNull(id, value, startDate, returnDate, isReturned);
+    public Loan(int id, float value, Date startDate, Date returnDate, boolean isReturned, Person assignee) {
```
Comment (marcus-ny) [inline src/main/java/seedu/address/model/person/Loan.java:56]: Do we need a constructor that indicates a `isReturned` status?

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -44,7 +43,8 @@ public ModelManager(ReadOnlyAddressBook addressBook, ReadOnlyUserPrefs userPrefs
         this.addressBook = new AddressBook(addressBook);
         this.userPrefs = new UserPrefs(userPrefs);
         filteredPersons = new FilteredList<>(this.addressBook.getPersonList());
-        currLoanList = FXCollections.observableList(new ArrayList<>());
+        filteredLoans = new FilteredList<>(this.addressBook.getLoanList());
+        sortedLoans = new SortedList<>(filteredLoans, Loan::compareTo);
```
Comment (marcus-ny) [inline src/main/java/seedu/address/model/ModelManager.java:47]: Making use of the behaviour of `SortedList` is really nice here. Removes the need for implicit sorting through functions at every change. 👍

File: src/main/java/seedu/address/logic/commands/ViewLoansCommand.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    public static final String COMMAND_WORD = "viewloans";
+
+    public static final String MESSAGE_SUCCESS = "Listed all loans";
+
+
+    @Override
+    public CommandResult execute(Model model) {
+        requireNonNull(model);
+        model.updateFilteredLoanList(PREDICATE_SHOW_ALL_ACTIVE_LOANS);
+        return new CommandResult(MESSAGE_SUCCESS, false, false, true);
```
Comment (marcus-ny) [inline src/main/java/seedu/address/logic/commands/ViewLoansCommand.java:21]:
```suggestion
        model.updateFilteredPersonList(unused -> false);
        return new CommandResult(MESSAGE_SUCCESS, false, false, true);
```

Comment (marcus-ny) [reply src/main/java/seedu/address/logic/commands/ViewLoansCommand.java:21]: This is just for cosmetic purposes and not really important. But when we call `viewloans` targeting all active loans, we should not show any particular contact's name at the top.

The UI will show a blank space at the top after this change. We will leave it to be fixed as an issue later on as this is cosmetic.

**Review by marcus-ny — CHANGES_REQUESTED** [review]
Changes look good to me. The app seems to behave as expected.

Just one or two minor nits to fix and should be gtg!

File: src/main/java/seedu/address/logic/commands/ViewLoansCommand.java (hunk 4 above)
Comment (narwhalsilent) [reply src/main/java/seedu/address/logic/commands/ViewLoansCommand.java:21]: I agree. In the "view all loans" GUI, the name of the loanee should also be shown for each loan,

File: src/main/java/seedu/address/model/person/Loan.java (hunk 2 above)
Comment (narwhalsilent) [reply src/main/java/seedu/address/model/person/Loan.java:56]: It may be useful for reading from the storage file.

**Review by marcus-ny — APPROVED** [review]

**Pull request merged**